# Homework 6: Hypothesis Testing, P-Hacking, and Simple Linear Regression 
***

**Name**: Matthew Sredojevic

***

This assignment is due on Moodle by **5pm on Friday April 20th**. Your solutions to theoretical questions should be done in Markdown/MathJax directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  Remember that you are encouraged to discuss the problems with your instructors and classmates, but **you must write all code and solutions on your own**.  For a refresher on the course **Collaboration Policy** click [here](https://github.com/dblarremore/csci3022/blob/master/resources/syllabus.md#collaboration-policy).

**NOTES**: 

- Do **NOT** load or use any Python packages that are not available in Anaconda 3.6. 
- Some problems with code may be autograded.  If we provide a function API **do not** change it.  If we do not provide a function API then you're free to structure your code however you like. 
- Because you can technically evaluate notebook cells is a non-linear order, it's a good idea to do Cell $\rightarrow$ Run All as a check before submitting your solutions.  That way if we need to run your code you will know that it will work as expected. 
- Submit only this Jupyter notebook to Moodle.  Do not compress it using tar, rar, zip, etc. 

In [1]:
import numpy as np
import matplotlib.pylab as plt
import pandas as pd
from scipy import stats 
%matplotlib inline

### [15 points] Problem 1 - Hypothesis Testing: Find-the-Knowledge-Bug 
***

As part of your job as a seasoned data science consultant, companies often bring you in to supervise their less-experienced data science teams before new product roll-outs.  On one such occasion, you are hired by a medium-sized internet-sales company that is preparing to release a new line of smart-home products.  Prior to their product release the company wants to do a targeted ad campaign to drive traffic to their site on launch day. 

You are asked to pair-up and do some inference work with a new employee named Stevey McKnowsNoStats. At various points in your day you catch Stevey making the following mistakes.  In each case, clearly explain to Stevey why his testing setup or conclusion is incorrect. 

**Part A**: Stevey has pulled up some data on the characteristics of customers that visited the company's website over the previous month.  He wants to perform an analysis on the mean age of customers that visit the site.  Let $X$ be the random variable describing the age of a site visitor and suppose that the population mean for $X$ is $\mu$. In particular, Stevey wants to see if the data suggests that the mean age of their customers is under 30 years old.   He decides to perform the test with a null hypothesis of $H_0: \bar{x} = 30$. 

---

Stevey, you idiot, you're using the wrong average. Youre using the population mean, where instead you should be using the dataset mean. instead of $\bar{x}$, you should be using $\mu$

**Part B**: After the wonderful explanation you gave him after the previous debacle, Stevey has seen the error in his ways and decides instead to do his hypothesis test with a null hypothesis of $H_0: \mu < 30$.  

No, Stevey. The null hypothesis needs to be what we assume from the data initailly. it shouldn't be  $\mu < 30$, but intead we'll assume that the average IS 30. Your Null hypothesis should read as $H_0: \mu = 30$

**Part C**: Finally on track with reasonable hypotheses of $H_0: \mu = 30$ and $H_1: \mu < 30$, Stevey computes a normalized test-statistic of $z = -1.35$ for the mean age and concludes that since $z = -1.35 < 0.05$ there is sufficient statistical evidence at the $\alpha = 0.05$ significance level that the mean age of their customers is less than 30.  

Oh, Stevey, you do not give me hope for humanity. $\alpha$ isn't the value you compare to your z value. You first need to find the z-value for $\alpha = 0.05$($z_{\alpha}$) and then compare those values

**Part D**: This time, with the hypotheses again $H_0: \mu = 30$ and $H_1: \mu < 30$, Stevey computes a p-value of $0.03$, and thus concludes that there is only a 3% probability that the null hypothesis is true. 

Well, the p-value you calculated, Im ASSUMING youre comparing it to the previous $\alpha = 0.05$ and since the p-value is smaler than $\alpha$ so you can REJECT the Null Hypothesis. We know at LEAST that the null hypothesis is not true because we were able to reject it.

### [10 points] Problem 2 - Scientists vs. P-Values 
***

Read the following article from **FiveThirtyEight**: [Statisticians Found One Thing They Can Agree On: It's Time to Stop Misusing P-Values](http://fivethirtyeight.com/features/statisticians-found-one-thing-they-can-agree-on-its-time-to-stop-misusing-p-values/). In what ways are scientists misusing p-values?  What suggestions are being made to use them properly? 

The p-value is being used to confirm hypothesis' that people are trying to prove, where instead the p-value is actually showing an accuracy of the data to the null hypothesis

Some of the suggestions are to rely less of p-values overall and instead use other tests 'such as confidence intervals or Bayesian analyses.', where as others believe that will lead to the same issue with these different tests. 

### [20 points] Problem 3 - Evaluating Pizza Delivery Performance 
***

The manager of a pizza chain with multiple locations likes to keep meticulous data on his pizza deliveries.  The data from more than 1200 deliveries in May 2014 is stored in pizza.csv. 

In [7]:
np.random.seed(1235)
dfP = pd.read_csv("data/pizza.csv")

In [8]:
dfP.head(20)

day       date       time operator  branch     driver  temperature  \
0   Thursday  01-May-14  33.708636    Laura    East      Bruno    71.433084   
1   Thursday  01-May-14  29.382070  Melissa    East  Salvatore    64.952920   
2   Thursday  01-May-14  33.580664  Melissa    West  Salvatore    49.113452   
3   Thursday  01-May-14  32.505369  Melissa    East  Salvatore    64.872559   
4   Thursday  01-May-14  25.493613  Melissa    West  Salvatore    59.630052   
5   Thursday  01-May-14  21.727466  Melissa  Centre      Bruno    53.715391   
6   Thursday  01-May-14  38.394169    Laura    West      Bruno    60.266862   
7   Thursday  01-May-14  22.186179  Melissa    West      Mario    61.187654   
8   Thursday  01-May-14  35.324994    Laura  Centre      Mario    64.204093   
9   Thursday  01-May-14  34.015280  Melissa  Centre      Bruno    67.383452   
10  Thursday  01-May-14  33.792085  Melissa  Centre      Bruno    70.983051   
11  Thursday  01-May-14  29.131207    Laura    East      Mario    69.949241   
12  Thursday  01-May-14  21.926461  Melissa    West      Mario    60.798941   
13  Thursday  01-May-14  28.185400  Melissa  Centre  Salvatore    72.440348   
14  Thursday  01-May-14  32.603202    Laura    West      Bruno    63.111336   
15  Thursday  01-May-14  32.648734  Melissa    West  Salvatore    55.732293   
16  Thursday  01-May-14  37.398504  Melissa  Centre      Mario    54.507040   
17  Thursday  01-May-14  33.956136  Melissa    East      Bruno    58.103934   
18  Thursday  01-May-14  26.379708    Laura    East      Mario    56.517764   
19  Thursday  01-May-14  33.193288  Melissa  Centre      Bruno    48.759259   

    bill  pizzas  
0   58.4       4  
1   26.4       2  
2   58.1       3  
3   35.2       3  
4   38.4       2  
5   61.8       4  
6   57.9       3  
7   35.8       2  
8   36.6       2  
9   44.8       5  
10  49.5       4  
11  49.7       3  
12  41.1       1  
13  42.9       1  
14  25.9       1  
15  41.5       6  
16  34.9       3  
17  42.4       4  
18  27.3       1  
19  47.7       4

**Part A**: The manager's goal is to have an average delivery time across all branches of less than 30 minutes.  Perform an appropriate hypothesis test at the $\alpha=0.05$ significance level to evaluate whether this goal has been achieved.  Be sure to clearly state your null and alternate hypothesis, describe your testing procedure, and show all calculations in Python. 

thanks stevey

$H_0: \mu = 30$

$H_1: \mu < 30$

the testing process is a z-test where we compare the actual mean to the assummed mean(the actual mean delivert time to the restriction of 30 minutes) and divide by the quantity standard deveation over the square root of the sample size (1266).

$$\frac{\bar{x} - \mu}{\frac{std}{\sqrt{n}}}$$

Then we compare this calculated value to out $z_{\alpha}$

In [59]:
x_barTime = 30
sigmaTime = dfP['time'].std()
n=len(dfP)
meanTime = dfP['time'].mean()
z_score = stats.norm.ppf(.05)

calcTime = (meanTime-x_barTime)/(sigmaTime/np.sqrt(n))

if z_score > calc:
    print("We cannot reject the Null Hypothesis")
else:
    print("We reject the Null Hypothesis")
print("Z test result:", calcTime)
print("Z_alpha score:", z_score)

We cannot reject the Null Hypothesis
Z test result: -3.17120774749
Z_alpha score: -1.64485362695


**Part B**: The manager also has the goal of having an average delivery temperature across all branches of greater than 65F.  Perform an appropriate hypothesis test at the $\alpha=0.05$ significance level to evaluate whether this goal has been achieved.  Be sure to clearly state your null and alternate hypothesis, describe your testing procedure, and show all calculations in Python. 

I used the same test as part A. The only thing that changes here are the standard deveation and mean and true mean. Everything else should be the same. Oh and the side of the test, this is a right tail test as opposed to the left tail test because this is a comparision greater than a given value (65 degrees)

$H_0: \mu = 65$

$H_1: \mu > 65$

In [73]:
x_barTemp = 65
sigmaTemp = dfP['temperature'].std()
meanTemp = dfP['temperature'].mean()

calcTemp = (meanTemp-x_barTemp)/(sigmaTemp/np.sqrt(n))

z_score = stats.norm.ppf(.95)

if z_score > calc:
    print("We cannot reject the Null Hypothesis")
else:
    print("We reject the Null Hypothesis")
print("Z test result:", calcTemp)
print("Z_alpha score:", z_score)

We cannot reject the Null Hypothesis
Z test result: 1.3820886485
Z_alpha score: 1.64485362695


**Part C**: The manager has enough money to give one of his delivery drivers a raise based on good performance. He knows from the data that his two best drivers are Mario and Luigi, but he's not sure if there is evidence that either one is better than the other.  Using delivery time and temperature as the criteria, perform any necessary hypothesis tests at the $\alpha = 0.05$ significance level, to determine if one of the drivers deserves a raise. Be sure to clearly describe your hypotheses and methodology, and show any relevant computations in Python. 

$H_0: \mu_M - \mu_L = 0$

$H_1: \mu_M - \mu_L > 0$

In [138]:
marioMeanTemp = dfP.loc[dfP['driver'] == "Mario", 'temperature'].mean()
marioSigmaTemp = dfP.loc[dfP['driver'] == "Mario", 'temperature'].std()
marioMeanTime = dfP.loc[dfP['driver'] == "Mario", 'time'].mean()
marioSigmaTime = dfP.loc[dfP['driver'] == "Mario", 'time'].std()
marioN = len(dfP.loc[dfP['driver'] == "Mario", 'temperature'])

luigiMeanTemp = dfP.loc[dfP['driver'] == "Luigi", 'temperature'].mean()
luigiSigmaTemp = dfP.loc[dfP['driver'] == "Luigi", 'temperature'].std()
luigiMeanTime = dfP.loc[dfP['driver'] == "Luigi", 'time'].mean()
luigiSigmaTime = dfP.loc[dfP['driver'] == "Luigi", 'time'].std()
luigiN = len(dfP.loc[dfP['driver'] == "Luigi", 'temperature'])

timeCalc = (marioMeanTime-luigiMeanTime)/np.sqrt(marioSigmaTime**2/(marioN)+luigiSigmaTime**2/(luigiN))
tempCalc = (marioMeanTemp-luigiMeanTemp)/np.sqrt(marioSigmaTemp**2/(marioN)+luigiSigmaTemp**2/(luigiN))  

print("Z-value for time:", timeCalc)
print("Z-value for temperature:", tempCalc)
print("Range:", stats.norm.ppf(.025), "to", stats.norm.ppf(.975))

Z-value for time: 0.00436618486748
Z-value for temperature: -0.956537156319
Range: -1.95996398454 to 1.95996398454


Since the z-values for both time and temperature are within the range of the 5% confidence interval, we can see that neither are significantly better. Therefore neither of them get the raise

**Part D**: On the other hand, the manager suspects that Salvatore has been underperforming as a driver.  The manager has a policy that if a pizza takes 40 minutes or longer to be delivered, he has to fork over a free bottle of wine.  This policy is only financially viable for the pizza chain if they have give out free wine on less than 5% of all pizza deliveries.  Perform a hypothesis test at the $\alpha = 0.05$ significance level to determine if there is significant evidence that the proportion of Salvatore's deliveries that take longer than 40 minutes is greater than 5%.  

$H_0: p = .05$<br\>
$H_1: p > .05$

In [108]:
salv = dfP.loc[dfP['driver'] == 'Salvatore']

$$z = \frac{\hat p - p}{\sigma}$$

$$\sigma^2 = \frac{p(1-p)}{n}$$

In [151]:
x_bar = 40

salvProp = len(salv.loc[salv['time']>40])/len(salv)
sigma = np.sqrt(.05*(1-.05)/(len(salv)))

z_score = (salvProp - .05)/sigma
print("Z score:",z_score)
print("Z alpha:", stats.norm.ppf(.95))

Z score: 0.445293626396
Z alpha: 1.64485362695


Since we can see that the z_score is below the z alpha for the .05 significance level, we can determine that we cannot reject the null hypothesis and so we cannot say that Salvador's deliveries take longer than 40 minutes greater tha 5% of the time.

### [15 points] Problem 4 - Naps vs Coffee for Memory? 
***

The consumption of coffee to benefit alertness is a common activity practiced by 90% of adults in North America. Often coffee is used to replace the need for sleep. One recent study compares students’ ability to recall memorized information after either the consumption of coffee or a brief nap. A random sample of 24 adults
were randomly divided into two groups and verbally given a list
of 25 words to memorize. During a break, one of the groups took a nap for an hour and a half,
another group stayed awake and were given a coffee an hour prior to testing. Researchers measured the number of words participants were
able to recall following the break. The summary statistics for the two groups are shown below.

$$
\begin{array}{c|c|c|c}
\textrm{Group} & \textrm{Sample Size} & \textrm{Mean} & \textrm{Standard Deviation} \\
\hline 
\textrm{Nap} & 12 & 15.5 & 3.2 \\ 
\textrm{Coffee} & 12 & 12.25 & 3.1 \\ 
\end{array}
$$

In previous experiments the mean recall measurement was found to be normally distributed. 

**Part A**: Compute a 95% t-confidence interval for the mean recall measurement for participants that took a nap before the test. Do all computations in Python. 

In [157]:
nap_x_bar = 15.5
nap_std = 3.2

mu = 25
n = 12

t_val = stats.t.ppf(.975, 11)
t_val
interval = nap_x_bar - t_val, nap_x_bar + t_val

print("T-confidence interval:", interval)

T-confidence interval: (13.299014839917051, 17.700985160082951)


**Part B**: Is there sufficient evidence, at the $\alpha = 0.05$ significance level, to conclude that taking a nap promotes better memory recall that drinking coffee?  Be sure to clearly explain the test that you're doing and state all hypotheses. Do all computations in Python. 

In [164]:
coffee_x_bar = 12.25
coffee_std = 3.1

z_score = (nap_x_bar - coffee_x_bar)/np.sqrt((nap_std**2/n)+coffee_std**2/n)
print("Z score:",z_score)
print("Range:", stats.norm.ppf(.025), "to", stats.norm.ppf(.975))


Z score: 2.52693300826
Range: -1.95996398454 to 1.95996398454


From the z score, we can see that there is a significant difference between naps and coffee because the 2.547>1.96 so we can concluse that there is a significant difference in favor of naps.

**Part C**: Compute a 95% confidence interval for the standard deviation of memory recall for coffee drinkers. Do all computations in Python.  

Using a Chi-Squared Distribution, we will find the 95% confidence interval for the standard devation with the formula

$$\frac{(n-1)s^2}{X^2_{\alpha/2, n-1}}$$

and

$$\frac{(n-1)s^2}{X^2_{1-\alpha/2, n-1}}$$

In [175]:
left = np.sqrt((11*coffee_std**2)/stats.chi2.ppf(1-.05/2, 11))
right = np.sqrt((11*coffee_std**2)/stats.chi2.ppf(.05/2, 11))

print("Confidence interval:",left,"to",right)

Confidence interval: 2.19602496767 to 5.26342185494


### [20 points] Problem 5 - Bad Science for Fun and Profit 
***

[Data Dredging](https://en.wikipedia.org/wiki/Data_dredging) and [p-hacking](https://www.explainxkcd.com/wiki/index.php/882:_Significant) are umbrella terms for the dangerous practice of automatically testing a large number of hypotheses on the entirety or subsets of a single dataset in order to find statistically significant results. In this exercise we will focus on the idea of testing hypotheses on subsets of a single data set.  

Johnny Nefarious has landed his first data science internship at an online marketing firm.  His primary summer project has been to design and test a new email advertisement for his company's best-selling product. To test his advertisement his supervisors have allowed him to send his ad to 4 targeted customer groups of 50 people every day for a month. 

The effectiveness of online advertising is typically measured by the ad's [click-through rate](https://en.wikipedia.org/wiki/Click-through_rate) (CTR), which is defined to be the _proportion_ of users that click on an advertisement. The company's standard email advertisement has a CTR of $0.05$.  Johnny is hoping to land a permanent position at the company when he graduates, so he's **really** motivated to show his supervisors that the CTR of his email advertisement is a (statistically) significant improvement over their previous ad. 

The data from Johnny's summer experiment can be found in email.csv. Load this dataset into Pandas. 

In [166]:
dfAd = pd.read_csv("data/email.csv")
dfAd.head()

Group         Date  Sent  Clicked
0      1          NaN    50        1
1      1  Oct  2 2017    50        2
2      1  Oct  3 2017    50        4
3      1          NaN    50        5
4      1  Oct  5 2017    50        1

**Part A**: State the null and alternate hypotheses that Johnny should test to see if his ad campaign is an improvement over the company's standard mailer with a CTR of $0.05$. 

$H_0: CTR = 0.05$

$H_1: CTR > 0.05$

**Part B**: Test the hypothesis from **Part A** at the $\alpha = 0.05$ significance level using a p-value test. Is there sufficient evidence for Johnny to conclude that his ad campaign is an improvement?   

In [190]:
length = len(dfAd)
summ = dfAd["Clicked"].sum()
mean = summ/(50*length)
z_score = (mean - 0.05)/np.sqrt(.05*(1-.05)/np.sum(dfAd["Sent"]))

p_val = 1-stats.norm.cdf(z_score)
print("P Value:", p_val)

P Value: 0.224365747475


**Part C**: Bummer, Johnny thinks. This is the part where he decides to resort to some questionable science.  Maybe there is a reasonable subset of the data for which his alternative hypothesis is supported?  Can he find it?  Can he come up for a reasonable justification for why this subset of the data should be considered while the rest should be discarded? 

Here are the **rules**: Johnny cannot modify the original data (e.g. by adding nonexistent clicks to certain groups or days) because his boss will surely notice.  Instead he needs to find a subset of the data for which his hypothesis is supported by a p-value test at the $\alpha = 0.05$ significance level _and_ be able to explain to his supervisors why his sub-selection of the data is reasonable.  

In addition to your explanation of why your successful subset of the data is potentially reasonable, be sure to thoroughly explain the details of the tests that you perform and show all of your Python computation.  

In [195]:
makeAdsGreatAgain = dfAd.loc[dfAd["Clicked"] > 0]

length = len(makeAdsGreatAgain)
summ = makeAdsGreatAgain["Clicked"].sum()
mean = summ/(50*length)
z_score = (mean - 0.05)/np.sqrt(.05*(1-.05)/np.sum(makeAdsGreatAgain["Sent"]))

p_val = 1-stats.norm.cdf(z_score)
print("P Value:", p_val)

P Value: 0.0336489140453


The reason why the initail dataset was inaccurate to show how well the ad functions works is because people who dont licks ads are using ad blockers (i mean dont we all in this day and age). so by ignorring all the people that ignore me, we can see that my formula is actually doing a great job because my p-value is less than than of the alpha value (.0336 < 0.5)

### [20 points] Problem 6 - Simple Linear Regression for Gas Mileage 
***

The data in auto.csv contains information on cars from the 1970s and 1980s. In this exercise you will construct a simple linear regression model for the response variable mpg with horsepower as the feature. Load the data into a Pandas DataFrame.    


SORRY

I just had a lot to do this week and was unable to make it to this last question. I honestly will look at this later for the sake of learning more about linear regression because this has actually popped up a lot in a few other classes this semester.

**0 POINTS TO SREDO-CLAW**

**30 POINTS TO DATASCI-FFENDOR**

In [196]:
dfAuto = pd.read_csv("data/auto.csv")
dfAuto.head()

model  year   mpg  cylinders  horsepower  weight
0  chevrolet chevelle malibu    70  18.0          8       130.0  3504.0
1          buick skylark 320    70  15.0          8       165.0  3693.0
2         plymouth satellite    70  18.0          8       150.0  3436.0
3              amc rebel sst    70  16.0          8       150.0  3433.0
4                ford torino    70  17.0          8       140.0  3449.0

**Part A**: Perform a simple linear regression with horsepower as the feature and mpg as the response.  Report the estimated regression model in the form $Y = \alpha + \beta x$. Do all computations in Python. 

**Part B**: Make a scatter-plot of the data with horsepower as the feature and mpg as the response and overlay the estimated regression line. 


**Part C**: Is the relationship between horsepower and mpg positive or negative? Justify your response. 

**Part D**: Find a 95% confidence interval for the slope parameter.  Based on this confidence interval, is there sufficient evidence to believe, at the 95% confidence level, that there is a real relationship between horsepower and mpg? 

**Part E**: Double-check your result from **Part D** by performing an equivalent hypothesis test.  Be sure to state your null and alternative hypotheses and explain your general methodology.  Do your conclusions agree with those made in **Part D**? 

**Part F**: What gas mileage would your simple linear regression model predict for a car with a horsepower of $97$? 